### PR TITLE
Ensure connection failures with pywbem 0.7.x are properly detected, closes #18

### LINKED ISSUE #18: Connection failures are not properly detected
Hi,
Because of this [try-except statement](https://github.com/Napsty/check_esxi_hardware/blob/89f9e505c398b382c5af0ba739efe9e6e5ea5cf7/check_esxi_hardware.py#L614),
when there is an issue during the initial connection, the code tries to call pywbem with different arguments and we end up with a non-meaningful exception:
```
TypeError: __init__() got an unexpected keyword argument 'no_verification'
```
I've never been able to test one of these patched pywbem versions but maybe:
* you could just catch SSL connection errors and have another except statement for other connection errors?
* you could look for the keyword in `pywbem.WBEMConnection.__init__.__code__.co_varnames` or using `inspect`?

### PR DESCRIPTION
**Pull request by peternewman**

### CODE REVIEW

**Comment by peternewman** [comment]
Untested since https://github.com/Napsty/check_esxi_hardware/issues/18#issuecomment-276050574

**Comment by pascalsaul** [comment]
$ ./check_esxi_hardware.py  -H <host> -U root -P <password>
OK - Server: Dell Inc. PowerEdge R710 s/n: <tag> System BIOS: 6.6.0 2018-05-22
$ ./check_esxi_hardware.py  -H <host> -U root -P <password>
OK - Server: Dell Inc. PowerEdge R740xd s/n: <tag> System BIOS: 2.2.11 2019-06-13

File: check_esxi_hardware.py [hunk 1]
```diff
@@ -659,10 +659,17 @@ def handler(signum, frame):
   try:
     conntest = pywbem.WBEMConnection(hosturl, (user,password))
     c = conntest.EnumerateInstances('CIM_Card')
-  except:
+  except pywbem.cim_http.Error, arg:
     #raise
```
Comment (sbraz) [inline check_esxi_hardware.py]: This comment should be simply removed

File: check_esxi_hardware.py [hunk 2]
```diff
@@ -659,10 +659,17 @@ def handler(signum, frame):
...
-    verboseoutput("Connection error, disable SSL certificate verification (probably patched pywbem)")
-    wbemclient = pywbem.WBEMConnection(hosturl, (user,password), no_verification=True)
+    if ( arg.__str__().find('SSL error: certificate verify failed') >= 0 ):
+      verboseoutput("Connection error, disable SSL certification verification (probably patched pywbem)")
+      wbemclient = pywbem.WBEMConnection(hosturl, (user,password), no_verification=True)
+    else:
+      verboseoutput("Other pywbem.cim_http.Error, continue as normal")
+      wbemclient = pywbem.WBEMConnection(hosturl, (user,password))
+  except:
```
Comment (sbraz) [inline check_esxi_hardware.py]: With v0.7.0, SSL errors are `pywbem.cim_operations.CIMError`s which are not subclasses of `pywbem.cim_http.Error` so they don't match the first `except`.

Comment (Napsty) [reply check_esxi_hardware.py]: Yep, agree. Not only in 0.7.0 but also later (tested here with 0.8.0-dev), but irrelevant in this if clause.

File: check_esxi_hardware.py (hunk 1 above)
Comment (Napsty) [reply check_esxi_hardware.py]: You refer to `#raise` ? I'd like to keep it as it's a handy way to identify (potentially) unknown cim errors.

**Comment by Napsty** [comment]
Thanks all you guys for the PR and the reviews. 
The except error handling case should be slightly changed as suggested by @sbraz because the certificate handling is handled by cim_operations not cim_http. 

```
# python
Python 2.7.12 (default, Nov 12 2018, 14:36:49)
[GCC 5.4.0 20160609] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import pywbem
>>> wbemclient = pywbem.WBEMConnection('https://esxi', ('root', 'waytoosecret'))
>>> c=wbemclient.EnumerateInstances('CIM_Card')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/dist-packages/pywbem/cim_operations.py", line 427, in EnumerateInstances
    **params)
  File "/usr/lib/python2.7/dist-packages/pywbem/cim_operations.py", line 188, in imethodcall
    raise CIMError(0, str(arg))
pywbem.cim_operations.CIMError: (0, 'SSL error: certificate verify failed')
```

Otherwise lgtm. We shouldn't put too much effort into this as pywbem 0.7.0 should meanwhile (hopefully) not be used anymore. Maybe some older legacy systems.

Debian: Starting from 8.0 (Jessie), newer 0.8.x was used  :heavy_check_mark: 
Ubuntu: Starting from 16.04 (Xenial), newer 0.8.x was used  :heavy_check_mark: 
Centos/RHEL 7: Use pywbem 0.7.0 :x: 
Centos/RHEL 8: Use pywbem 0.11.x  :heavy_check_mark: 
SUSE SLES: At least since SLES 11 using 0.9.x (in additional repo) :heavy_check_mark:

So we're mostly talking about Centos/RHEL 7 systems here - however they will still stuck around for a couple of years.

**Comment by sbraz** [comment]
> So we're mostly talking about Centos/RHEL 7 systems here - however they will still stuck around for a couple of years.

Yep that's the reason I'm running 0.7 :/

**Comment by Napsty** [comment]
@sbraz I stumbled on https://www.zenetys.com/run-check-esxi-hardware-py-on-el6-with-pywbem-0-11-0/ recently. Did you already give this a shot?

@peternewman Were you able to adjust the code according to the discussion and update the PR? Will gladly merge when done, thanks!

**Comment by sbraz** [comment]
> @sbraz I stumbled on https://www.zenetys.com/run-check-esxi-hardware-py-on-el6-with-pywbem-0-11-0/ recently. Did you already give this a shot?

I'd rather not try installing older versions or anything like that. It would be nice to see this PR fixed for 0.7 instead.

**Comment by peternewman** [comment]
> @peternewman Were you able to adjust the code according to the discussion and update the PR? Will gladly merge when done, thanks!

Sorry for the delay. There is also a route to generate the error I was trapping, see here pywbem.cim_http.Error:
https://github.com/pywbem/pywbem/blob/ab9ef9a332e7e32d02c7cc2861ab0f6e2c05d04c/trunk/cim_http.py#L140-L141

I can't reproduce it now, but that must have been going at one point for me to put the code in.

Are you happy if I add the other except too?

**Comment by Napsty** [comment]
Sorry @peternewman , thought I already responded here a while ago.
Yes, please add the other except, too or use a general "except" without specifying the pywbem function (maybe even easier this way?)

**Comment by Napsty** [comment]
As discussed in the related issue #18, cannot reproduce with current CentOS 7.8 and meanwhile newer pywbem versions are available and should be used.

**Pull request closed without merging**